feat: spa option, `preview` and `dev` for MPA and SSR apps

brillout opened the pull request:
### Description

This fixes `$ vite preview` and `$ vite dev` for SSR and MPA apps.

This PR enables users and frameworks to set `config.isSPA = false` to disable SPA catch-all routing.

> The docs added by this PR include a description of what "SPA catch-all routing" means.

It supersedes these previous PRs:
 - https://github.com/vitejs/vite/pull/4640
 - #7665
 - https://github.com/vitejs/vite/pull/8061

And fixes these issues:
 - https://github.com/vitejs/vite/issues/7631
 - https://github.com/vitejs/vite/issues/5720

### Additional context

I decided against `config.appType: spa | ssr | mpa` because many frameworks blur the line between `ssr` and `mpa`. For example, vite-plugin-ssr 0.4 allows the user to implement SSR as well as MPA apps. It also allows the user to have one SSR page while the rest is MPA; it's then not clear whether the app type is `ssr` or `mpa`. Many other frameworks, such as Nuxt, also blur the line between SSR and MPA.

Thus I believe having only two modes "SPA" or "not SPA" is the way to go.

I'm fairly confident this PR is in a good state ~~and can be merged as-is~~ on a high-level (edit). (But happy to be shown wrong.)

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

brillout committed:
fix: `$ vite preview` and `$ vite dev` for MPA and SSR apps

brillout committed:
chore: minor code readability improvement

benmccann commented on docs/config/index.md:
```diff
@@ -448,6 +448,16 @@ export default defineConfig(({ command, mode }) => {
+- **Type:** `boolean`
+- **Default:** `true`
+
+  This toggles the SPA catch-all routing of the development and preview
+  server: by default, all URLs are rendered to the same single page (in order
+  to enable SPA client-side routing). For SSR and MPA apps, this behavior can
+  be disabled by setting `isSPA` to `false`.
```
the other lines in this file are not manually wrapped
```suggestion
  This toggles the SPA catch-all routing of the development and preview server: by default, all URLs are rendered to the same single page (in order to enable SPA client-side routing). For SSR and MPA apps, this behavior can be disabled by setting `isSPA` to `false`.
```

benmccann commented on packages/vite/src/node/config.ts:
```diff
@@ -205,6 +205,17 @@ export interface UserConfig {
     >
   }
+  /**
+   * Whether the app is an SPA.
+   *
+   * This toggles the SPA catch-all routing of the development and preview
+   * server: by default, all URLs are rendered to the same single page (in order
+   * to enable SPA client-side routing). For SSR and MPA apps, this behavior can
+   * be disabled by setting `isSPA` to `false`.
+   *
+   * @default true
```
the other options don't have a new line before `@default`
```suggestion
   * Whether the app is a SPA.
   * This toggles the SPA catch-all routing of the development and preview
   * server: by default, all URLs are rendered to the same single page (in order
   * to enable SPA client-side routing). For SSR and MPA apps, this behavior can
   * be disabled by setting `isSPA` to `false`.
   * @default true
```

brillout committed:
chore: update docs/config/index.md

Co-authored-by: Ben McCann <322311+benmccann@users.noreply.github.com>

brillout committed:
chore: update packages/vite/src/node/config.ts

Co-authored-by: Ben McCann <322311+benmccann@users.noreply.github.com>

benmccann commented on packages/vite/src/node/config.ts:
```diff
@@ -285,6 +294,7 @@ export async function resolveConfig(
   defaultMode = 'development'
 ): Promise<ResolvedConfig> {
   let config = inlineConfig
+  config.isSPA ??= true
```
it seems like the other config defaults are specified around line 480. is there a reason to do this one here instead of down there?

brillout replied:
👍 & done. (It also makes a functional difference: the default is set after the `config()` hooks are executed which is more consistent with the rest of the code. Good catch, thanks 👌.)

brillout committed:
fix: move `config.isSPA ??= true` after `config()` hooks execution

benmccann commented on packages/vite/src/node/server/index.ts:
```diff
@@ -503,20 +503,25 @@ export async function createServer(
-  // run post config hooks
-  // This is applied before the html middleware so that user middleware can
-  // serve custom content instead of index.html.
+  // We need to apply post server hooks before `spaFallbackMiddleware()`.
```
I tried to make this change in an earlier PR but @patak-dev reported that it broke Vitepress: https://github.com/vitejs/vite/pull/4670#issuecomment-907748589. Perhaps it's okay to change in a major version though? @patak-dev would probably know better as I'm not familiar with Vitepress

brillout replied:
Made a Vitepress PR to fix the situation: https://github.com/vuejs/vitepress/pull/625.

benmccann commented on packages/vite/src/node/config.ts:
```diff
@@ -469,6 +478,8 @@ export async function resolveConfig(
 
   const optimizeDeps = config.optimizeDeps || {}
 
+  config.isSPA ??= true
```
I was thinking that you'd have something like `isSPA: true,` below in the creation of `resolved` and then you'd use `resolved` when checking the value of `isSPA`

brillout replied:
Actually, I've just pushed a commit to resolve `isSPA`'s default value at the very end, just before it's used.

This provides more flexibility: it allows the framework to overwrite Vite's default value while enabling the user to overwrite the framework provided default value.

Let me elaborate.

For example, this is how vite-plugin-ssr and Telefunc are ensuring that the server and preview default port is `3000` for Vite 3 (despites the fact that Vite 3 changed the default port from `3000` to `5173` which I don't like as I find it difficult to remember):

https://github.com/vikejs/telefunc/blob/62d2753fe452e3bb5981da110b7e1170b35b1c61/telefunc/node/vite/plugins/previewConfig.ts#L23-L27

The neat thing here is that user can still set a different port.

That works because Vite resolves the default value of `config.server.port` and `config.preview.port` at the very end:

https://github.com/vitejs/vite/blob/e37a51dc767fe3fcf2eb49f7f76eab63d90f777a/packages/vite/src/node/server/index.ts#L572-L580
https://github.com/vitejs/vite/blob/e37a51dc767fe3fcf2eb49f7f76eab63d90f777a/packages/vite/src/node/preview.ts#L116-L122

I'm marking this conversation as resolved but let me know if you disagree and I'll unresolve it.

brillout committed:
fix: resolve `isSPA`'s default at very end

brillout committed:
fix: typo

brillout committed:
fix: improve comment accuracy

benmccann commented on packages/vite/src/node/server/index.ts:
```diff
@@ -503,21 +503,28 @@ export async function createServer(
-  if (!middlewareMode || middlewareMode === 'html') {
+  const isMiddlewareModeSSR = middlewareMode && middlewareMode !== 'html'
+
+  const isSPA = config.isSPA ?? true
```
This still seems like a weird way to handle the default value to me. It means the default has to be specified anywhere we'd want to use `isSPA`

SvelteKit does the same thing you mentioned where we provide a framework default for some config values and allow the user to override them. It works just fine with traditionally specified config values, so I'm not sure why we'd have to do something special for this one

patak-dev replied:
`isSPA` should be resolved in [resolveConfig](https://github.com/brillout/vite/blob/5de9eb5ba32c2c3f25bbd937c35bd99aaba0fab9/packages/vite/src/node/config.ts#L292), so there is no need for this as @benmccann said.

brillout replied:
Done.

benmccann commented on packages/vite/src/node/server/index.ts:
```diff
@@ -503,21 +503,28 @@ export async function createServer(
+  if (isSPA && !isMiddlewareModeSSR) {
+    // SPA catch-all fallback routing
+    middlewares.use(spaFallbackMiddleware(root))
     // transform index.html
```
We should figure out what we want to do about the `transformIndexHtml` plugin hook (one of the things I mentioned in https://github.com/vitejs/vite/issues/7631).

At the very least we should probably update the documentation at https://vitejs.dev/guide/api-plugin.html#transformindexhtml clarifying the conditions under which this plugin hook is invoked especially now that it lives inside this new `if` block.

- Do we want to say that it's not supported during SSR?
- Do we want to figure out how to ensure it's called even when doing SSR? E.g. we could create a spec that says SSR frameworks must call `server.transformIndexHtml` during dev on each request for the template and that Vite will invoke it during the build process. But the functionality would be somewhat limited as it would only be called on the template and not the final rendered page (or maybe could also be called for prerendered pages if we somehow figured out how to get prerendering to happen in a Vite plugin). E.g. SvelteKit currently has a file at `src/app.html` which has placeholders like `%svelte.head` and `%svelte.body%` that are replaced with the dynamically generated content. We can't do much transform since there are no contents present (including even links to JS files). The only usecase I've heard of is someone who wanted to look for a favicon and replace it with multiple versions including retina support.
- I also sort of wonder if this plugin hook even makes sense. It was designed for SPA mode and I'm not sure it's something we'd have come up with if thinking about SSR. Is it heavily used in SPA apps and would it be missed if we were to remove it instead?

patak-dev replied:
We talked in today's meeting about `transformIndexHtml`. We should document that this hook isn't called during SSR.

brillout replied:
Yes I agree about `transformIndexHtml()`. Although note that vite-plugin-ssr does need `transformIndexHtml()` because it injects different HMR code depending on whether the users uses Vue, Svelte, React, etc. I made a comment about it a while ago over there https://github.com/vitejs/vite/pull/7993#issuecomment-1116040446.

I think we can proceed with this PR independently of the `transformIndexHtml()` discussion.

As for documenting, see my comment below.

patak-dev commented:
We talked about the PR in today's team meeting. We're worried about making Vite preview ever more complex to accommodate future requirements from users like vite-plugin-ssr. We still think that a `spa` option (I agree with Anthony's rename request) is a necessity given current hardcoded assumptions in Vite (independent of the preview changes).
So let's move forward with the PR in v3, but frameworks using vite preview instead of a wrapper needs to understand that future features need to be done in middleware. It is out of the scope of Vite to keep augmenting vite preview to support other requirements.

patak-dev commented on packages/vite/src/node/preview.ts:
```diff
@@ -97,19 +98,41 @@ export async function preview(
+  if (!isSPA) {
+    // We apply the plugins' server post hooks after `sirv()` so that `sirv()`
+    // can serve pre-rendered pages before any SSR middleware.
+    postHooks.forEach((fn) => fn && fn())
```
Maybe I'm missing some context but couldn't these hooks be always applied before sirv? And if every hook ignores pre-rendered pages then they will be served by sirv? Looks more flexible so even for pre-rendered pages the hooks could do some special handling. It means checking if the file exists in disk, no?
I like the idea of static assets / 404 handling being the last fallback after the postHooks in both modes.

brillout replied:
The thing is that SSR middlewares actually do not ignore pre-rendered pages: a pre-rendered page `/about` has been rendered at build-time to `dist/client/about.html` but if the user were to `$ rm ./dist/client/about.html` then the SSR middleware will pick up and render the page at run-time.

That's at least how vite-plugin-ssr works and I expect other frameworks to do this as well. (Because it's a canonical way to support partial pre-rendering.)

Thus, `sirv()` needs to run before the SSR middleware, otherwise the SSR middleware will wrongfully render the pre-rendered pages and `sirv()` never runs. This breaks `$ vite preview` in the sense that it's not a truthful representation of production anymore.

patak-dev replied:
Again, I may be missing some context, so thanks for the explanations. Why SSR middlewares can't check if the file exists before deciding what to do? Doesn't something similar would happen in the server where you have a chance to intercept a request and do whatever you want (even if it is pre-rendered, you may want to ignore it depending on some other variables?). @benmccann @antfu maybe you could help here, I'm just looking to avoid cornering us into fighting against sirv later. We aren't going to be able to add a pre/post middleware strategy here, so good to get this one right.

benmccann replied:
Previously, the post hooks were always being run after sirv. You're talking about the SSR case, but it seems the case that has changed is the SPA case and the hooks are now being run before sirv in that case. Why does the SPA case need to change? I see the comment explaining it needs to happen first because `single` will catch everything, but then I don't understand why it's happening after in `main` today. Is `main` broken? Or am I reading the code incorrectly? (very possible as it's time for me to go to bed...)

brillout replied:
Turns out my Vitepress PR (https://github.com/vuejs/vitepress/pull/625) doesn't work for URLs ending with `.html`. I'm not sure how to fix Vitepress so I decided to revert all changes related to SPA mode.

The SPA mode of this PR now behaves exactly as `main`.

As you can see the PR now changes almost nothing: it's mostly about setting `single` to `false` for SSR (this change was already approved & merged before but got reverted because it needs to be applied only in SSR mode). The rest of the changes are actually optional, but I guess they make sense (e.g. Ben's point that some SPA middlewares are superfluous for SSR).

So I guess it's ready for merge, finally :-).

As for SSR, let me elaborate.

In general, apps are usually served like this:
 - First the URL is checked against the Static Host.
 - If the Static Host doesn't return a response, then forward the HTTP request to the Node.js server.

That's how reverse proxies e.g. NGNIX are set up. It's also how the routing part of Cloudflare workers are written. This seems standard.

Translated to this PR:
 - First `sirv()`
 - If `sirv()` doesn't deliver an HTTP response, then the SSR middleware.

In other words, the `sirv()` middleware comes first, then post hooks middlewares.

brillout replied:
I'm confident all concerns are addressed; I'm marking this as resolved.

brillout committed:
fix: rename `isSPA` to `spa`

brillout committed:
fix: resolve default value in `resolveConfig()`

brillout committed:
docs: refactor documentation strategy

brillout committed:
docs: add draft text

brillout committed:
docs: remove draft text

brillout commented:
I've updated the PR and I believe it resolves all discussions.

I've refactored how we document this: the documentation now lives under `Guide - SSR` which I believe to be a more natural fit.

> We talked in today's meeting about `transformIndexHtml`. We should document that this hook isn't called during SSR.

This is somewhat documented at https://vitejs.dev/guide/ssr.html#setting-up-the-dev-server and since the docs now live on that same page, I believe it's enough.

> We're worried about making Vite preview ever more complex to accommodate future requirements from users [...] but frameworks using vite preview instead of a wrapper needs to understand that future features need to be done in middleware

I agree and this is actually what this PR is about: getting Vite's internal SPA assumptions out of the way in order to uncouple Vite from user middlewares.

Telefunc now also adds its RPC middleware to the development and preview servers and it works. I believe that, after this PR, things will be flexible enough to accommodate for all kinds of needs. Some minor tweaks may be needed, but nothing fundamental AFAICT. And, yes, for unreasonable requirements we can answer: write your own CLI.

benmccann commented on docs/config/index.md:
```diff
@@ -462,6 +462,13 @@ export default defineConfig(({ command, mode }) => {
+- **Type:** `boolean`
+- **Default:** `true`
+
+  This configuration is meant for framework authors, or users using Vite's SSR API. See [SSR - Vite CLI](/guide/ssr#vite-cli).
```
I don't know if we need to mention framework authors. We should probably define SPA. I'm not sure why linking to the CLI instead of SSR page generally. If we were going to keep the existing text, the comma is unnecessary
```suggestion
  Whether your application is a Single Page Application (SPA). Set to `false` if using Vite's [SSR](/guide/ssr) functionality.
```

patak-dev replied:
This sounds a bit strange. VitePress is going to use the SSR functionality but keep `spa: true`, no? Maybe something like this?
```
  Whether your application is a Single Page Application (SPA). Set to `false` for other kinds of apps like MPAs. Learn more in Vite's [SSR guide](/guide/ssr#vite-cli).
```

brillout replied:
>   Whether your application is a Single Page Application (SPA). Set to `false` for other kinds of apps like MPAs. Learn more in Vite's [SSR guide](/guide/ssr#vite-cli).

Committed: https://github.com/vitejs/vite/pull/8217/commits/71187f318e993949737f782ea27e9224fd60583c.

brillout committed:
Update docs/config/index.md

Co-authored-by: Ben McCann <322311+benmccann@users.noreply.github.com>

brillout committed:
fix: udpate JSDoc

brillout committed:
fix: revert SPA mode while keeping SSR mode as-is

benmccann commented on packages/vite/src/node/preview.ts:
```diff
@@ -97,20 +98,35 @@ export async function preview(
   // apply post server hooks from plugins
   postHooks.forEach((fn) => fn && fn())
 
+  if (!spa) {
```
I'm not entirely sure we should add this block. It seems a bit too specific to a single use case. E.g. SvelteKit targets many different platforms and not all of them will have this behavior. I think it'd be something that would be easy to add in a middleware rather than adding it here

If we were to keep it, do we actually need the `if (!spa)` check here?

brillout replied:
Makes sense. Done.

brillout committed:
fix: remove 404 handler

benmccann commented on packages/vite/src/node/preview.ts:
```diff
@@ -1,5 +1,6 @@
 import path from 'path'
 import type * as http from 'http'
+import fs from 'fs'
```
Looks like this unused import needs to be removed for the CI to pass
```suggestion
```

brillout committed:
Update packages/vite/src/node/preview.ts

Co-authored-by: Ben McCann <322311+benmccann@users.noreply.github.com>

benmccann reviewed (dismissed):
Awesome! Thanks for driving this and seeing it through. I'm really excited about it!

brillout committed:
fix: apply patak doc suggestion

brillout commented:
Alright, I've applied all suggestions. Let me know if I missed something. Also feel free to directly push to the PR.

> Awesome! Thanks for driving this and seeing it through. I'm really excited about it!

I'm glad you're excited about this. Thank you both for your helpful reviews.

patak-cat committed:
chore: update

benmccann commented on docs/config/index.md:
```diff
@@ -462,6 +462,13 @@ export default defineConfig(({ command, mode }) => {
+- **Type:** `boolean`
+- **Default:** `true`
+
+  Whether your application is a Single Page Application (SPA). Set to `false` for other kinds of apps like MPAs. Learn more in Vite's [SSR guide](/guide/ssr#vite-cli).
```
Should this link have an `.html` ending? The link in the next `.md` file does, so I figure one of the two is probably wrong, but I'm not sure which one

brillout replied:
I don't know. @patak-dev ?

brillout replied:
(I've seen both in the docs, so I was assuming it doesn't matter. But I guess patak will know more about this.)

patak-dev replied:
It isn't needed 👍🏼

benmccann commented on docs/guide/ssr.md:
```diff
@@ -264,3 +264,14 @@ In some cases like `webworker` runtimes, you might want to bundle your SSR build
+
+The CLI commands `$ vite dev` and `$ vite preview` can also be used for SSR apps:
+
+1. Add your SSR middleware to the development server with [`configureServer`](/guide/api-plugin.html#configureserver) and to the preview server with [`configurePreviewServer`](/guide/api-plugin.html#configurepreviewserver).
```
I guess we should remove these then since the `.html` is needed per https://github.com/vitejs/vite/pull/8217#discussion_r878917097
```suggestion
1. Add your SSR middleware to the development server with [`configureServer`](/guide/api-plugin#configureserver) and to the preview server with [`configurePreviewServer`](/guide/api-plugin#configurepreviewserver).
```

patak-cat committed:
chore: update

Co-authored-by: Ben McCann <322311+benmccann@users.noreply.github.com>
